### ISSUE TITLE
Binance websocket orderbook returning inconsistent values across parallel runs

### ISSUE THREAD

**Issue by jsmerix** [ccxt/ccxt#21254]
### Operating System

Ubuntu

### Programming Languages

Python

### CCXT Version

4.2.45

### Description

Hello, 

I have found an inconsistency during Binance futures websocket orderbook download after running the following experiment:

1. Two standalone docker containers running Ubuntu collecting Binance websocket orderbook data for ~5 minutes kicked-off at the same time, running in parallel (completely isolated from each other)

Here is a simplified version of the CCXT code I used (I have removed non-CCXT related surrounding code because it would be confusing)

```
while True:
    try:
        await exchange.watch_order_book_for_symbols(
            symbols, limit=limit
        )
    except Exception as e:
            _LOG.info("Subscription to symbols failed %s", e)
            # Handling potential issues: Sometimes the order book may fall out of sync,
            # leading to an error like "handleOrderBook received an out-of-order nonce".
            # Sleeping for 1 second to allow the system to catch up and get back in sync.
            # ref. https://github.com/ccxt/ccxt/issues/17827#issuecomment-1537532598
            await exchange.sleep(1000)
    
     for pair in symbols:
        data = exchange.orderbooks[pair].limit()
        process_data(data)
```

2. After the downloaders finished I have analysed the data and found two data points (each generated by one container) where for a given timestamp the orderbooks did not match:

First data point:

```
{'bids': [[2810.04, 9.223],
  [2810.03, 11.557],
  [2810.01, 0.728],
  [2810.0, 3.67],
  [2809.99, 17.473],
  [2809.96, 0.355],
  [2809.94, 0.071],
  [2809.93, 0.001],
  [2809.91, 29.944],
  [2809.9, 19.93]],
 'asks': [[2810.08, 0.023],
  [2810.12, 0.023],
  [2810.15, 0.008],
  [2810.19, 0.008],
  [2810.2, 0.015],
  [2810.22, 0.074],
  [2810.25, 6.027],
  [2810.27, 0.356],
  [2810.29, 1.068],
  [2810.3, 27.383]],
 'timestamp': 1708074512151,
 'datetime': '2024-02-16T09:08:32.151Z',
 'nonce': 3988362550143
}
```

Second data point:
```
{'bids': [[2810.04, 9.223],
  [2810.03, 11.557],
  [2810.01, 0.728],
  [2810.0, 3.67],
  [2809.99, 17.473],
  [2809.96, 0.355],
  [2809.94, 0.071],
  [2809.93, 0.001],
  [2809.91, 29.944],
  [2809.9, 19.93]],
 'asks': [[2810.05, 0.043],
  [2810.08, 0.023],
  [2810.1, 3.008],
  [2810.12, 0.023],
  [2810.15, 0.008],
  [2810.16, 6.108],
  [2810.19, 0.008],
  [2810.2, 0.015],
  [2810.21, 0.035],
  [2810.22, 0.074]],
 'timestamp': 1708074512151,
 'datetime': '2024-02-16T09:08:32.151Z',
 'nonce': 3988362550143,
}
```

As we can observe, the `nonce` and `timestamp` are matching between the two data points but the lists in the `asks` field do not match. Furthermore, I have collected raw messages as they came from Binance during runtime of both downloaders. The raw message was identical for each of the data points:

```
{
'e': 'depthUpdate',
  'E': 1708074512151,
  'T': 1708074512150,
  's': 'ETHUSDT',
  'U': 3988362531055,
  'u': 3988362550143,
  'pu': 3988362530912,
  'b': [['500.00', '76.200'],
   ['1000.00', '105.000'],
   ['1966.95', '0.000'],
   ['2767.66', '0.000'],
   ['2767.67', '1.450'],
   ['2777.05', '0.000'],
   ['2781.73', '8.939'],
   ['2783.97', '0.730'],
   ['2796.47', '1.474'],
   ['2797.13', '0.000'],
   ['2797.39', '42.144'],
   ['2800.65', '0.689'],
   ['2801.31', '1.000'],
   ['2801.40', '47.913'],
   ['2802.37', '0.016'],
   ['2802.48', '3.741'],
   ['2802.87', '0.296'],
   ['2802.91', '41.008'],
   ['2803.22', '31.958'],
   ['2803.29', '79.980'],
   ['2803.31', '26.392'],
   ['2803.34', '25.366'],
   ['2803.85', '1.397'],
   ['2804.02', '1.646'],
   ['2804.05', '0.059'],
   ['2804.42', '0.040'],
   ['2804.51', '0.151'],
   ['2804.87', '1.602'],
   ['2805.29', '0.000'],
   ['2805.91', '1.627'],
   ['2806.04', '45.795'],
   ['2806.07', '0.000'],
   ['2806.23', '1.471'],
   ['2806.27', '0.365'],
   ['2806.36', '2.019'],
   ['2806.39', '45.799'],
   ['2806.42', '8.316'],
   ['2806.45', '0.127'],
   ['2806.46', '3.091'],
   ['2806.63', '29.185'],
   ['2806.84', '27.828'],
   ['2806.86', '0.164'],
   ['2806.94', '5.402'],
   ['2807.01', '0.150'],
   ['2807.06', '45.547'],
   ['2807.07', '0.270'],
   ['2807.12', '0.128'],
   ['2807.13', '0.207'],
   ['2807.18', '0.626'],
   ['2807.20', '3.300'],
   ['2807.23', '2.268'],
   ['2807.27', '0.564'],
   ['2807.30', '3.586'],
   ['2807.32', '2.142'],
   ['2807.33', '18.415'],
   ['2807.35', '0.584'],
   ['2807.37', '0.356'],
   ['2807.42', '0.016'],
   ['2807.50', '67.163'],
   ['2807.56', '0.041'],
   ['2807.62', '1.585'],
   ['2807.64', '22.614'],
   ['2807.68', '0.209'],
   ['2807.70', '27.061'],
   ['2807.76', '2.851'],
   ['2807.81', '20.623'],
   ['2807.83', '10.301'],
   ['2807.84', '13.560'],
   ['2807.86', '3.504'],
   ['2807.88', '0.160'],
   ['2807.89', '23.239'],
   ['2807.90', '0.181'],
   ['2807.92', '0.264'],
   ['2807.93', '0.440'],
   ['2807.94', '35.870'],
   ['2807.96', '2.892'],
   ['2807.98', '2.821'],
   ['2807.99', '1.227'],
   ['2808.02', '0.000'],
   ['2808.04', '0.015'],
   ['2808.06', '72.702'],
   ['2808.08', '17.800'],
   ['2808.09', '7.400'],
   ['2808.10', '22.335'],
   ['2808.12', '17.147'],
   ['2808.13', '1.236'],
   ['2808.14', '4.010'],
   ['2808.15', '18.949'],
   ['2808.16', '0.568'],
   ['2808.17', '8.166'],
   ['2808.19', '30.851'],
   ['2808.23', '0.540'],
   ['2808.24', '1.230'],
   ['2808.25', '4.717'],
   ['2808.26', '1.467'],
   ['2808.27', '0.282'],
   ['2808.29', '1.018'],
   ['2808.30', '0.696'],
   ['2808.31', '19.282'],
   ['2808.32', '35.153'],
   ['2808.34', '0.612'],
   ['2808.36', '0.455'],
   ['2808.37', '25.305'],
   ['2808.40', '25.040'],
   ['2808.41', '0.124'],
   ['2808.42', '24.342'],
   ['2808.43', '20.426'],
   ['2808.44', '0.255'],
   ['2808.45', '18.363'],
   ['2808.46', '17.086'],
   ['2808.47', '187.436'],
   ['2808.48', '23.126'],
   ['2808.49', '20.935'],
   ['2808.52', '56.576'],
   ['2808.53', '0.000'],
   ['2808.55', '37.888'],
   ['2808.56', '17.992'],
   ['2808.57', '16.459'],
   ['2808.58', '0.061'],
   ['2808.59', '4.646'],
   ['2808.60', '55.603'],
   ['2808.62', '21.963'],
   ['2808.63', '14.739'],
   ['2808.64', '2.087'],
   ['2808.65', '52.373'],
   ['2808.66', '19.916'],
   ['2808.67', '0.825'],
   ['2808.68', '17.790'],
   ['2808.69', '35.234'],
   ['2808.70', '2.865'],
   ['2808.71', '0.711'],
   ['2808.73', '9.115'],
   ['2808.74', '0.000'],
   ['2808.76', '11.312'],
   ['2808.77', '0.356'],
   ['2808.78', '4.788'],
   ['2808.81', '0.356'],
   ['2808.83', '7.039'],
   ['2808.85', '1.623'],
   ['2808.87', '0.712'],
   ['2808.88', '2.574'],
   ['2808.89', '2.276'],
   ['2808.91', '5.532'],
   ['2808.96', '27.691'],
   ['2808.97', '1.863'],
   ['2808.99', '14.634'],
   ['2809.00', '18.015'],
   ['2809.02', '9.831'],
   ['2809.03', '7.225'],
   ['2809.05', '9.540'],
   ['2809.06', '0.008'],
   ['2809.16', '0.468'],
   ['2809.17', '1.600'],
   ['2809.21', '0.020'],
   ['2809.22', '10.653'],
   ['2809.24', '4.891'],
   ['2809.27', '14.204'],
   ['2809.30', '4.675'],
   ['2809.33', '2.670'],
   ['2809.34', '5.117'],
   ['2809.35', '0.355'],
   ['2809.39', '8.382'],
   ['2809.42', '2.093'],
   ['2809.46', '2.261'],
   ['2809.47', '0.000'],
   ['2809.54', '0.798'],
   ['2809.56', '0.000'],
   ['2809.57', '0.100'],
   ['2809.59', '0.071'],
   ['2809.61', '0.279'],
   ['2809.62', '0.350'],
   ['2809.67', '1.598'],
   ['2809.68', '3.163'],
   ['2809.71', '3.618'],
   ['2809.75', '0.071'],
   ['2809.76', '3.568'],
   ['2809.78', '1.153'],
   ['2809.79', '0.071'],
   ['2809.80', '4.025'],
   ['2809.81', '2.313'],
   ['2809.82', '22.203'],
   ['2809.83', '35.280'],
   ['2809.84', '0.000'],
   ['2809.85', '0.000'],
   ['2809.86', '29.007'],
   ['2809.87', '8.993'],
   ['2809.88', '2.299'],
   ['2809.89', '0.000'],
   ['2809.90', '19.930'],
   ['2809.91', '29.944'],
   ['2809.92', '0.000'],
   ['2809.93', '0.001'],
   ['2809.94', '0.071'],
   ['2809.95', '0.000'],
   ['2809.96', '0.355'],
   ['2809.97', '0.000'],
   ['2809.98', '0.000'],
   ['2809.99', '17.473'],
   ['2810.00', '3.670'],
   ['2810.01', '0.728'],
   ['2810.02', '0.000'],
   ['2810.03', '11.557'],
   ['2810.04', '9.223']],
  'a': [['2809.83', '0.000'],
   ['2809.84', '0.000'],
   ['2809.86', '0.000'],
   ['2809.87', '0.000'],
   ['2809.88', '0.000'],
   ['2809.90', '0.000'],
   ['2809.91', '0.000'],
   ['2809.92', '0.000'],
   ['2809.94', '0.000'],
   ['2809.95', '0.000'],
   ['2809.96', '0.000'],
   ['2809.99', '0.000'],
   ['2810.00', '0.000'],
   ['2810.02', '0.000'],
   ['2810.03', '0.000'],
   ['2810.04', '0.000'],
   ['2810.06', '0.000'],
   ['2810.08', '0.023'],
   ['2810.09', '0.000'],
   ['2810.11', '0.000'],
   ['2810.12', '0.023'],
   ['2810.15', '0.008'],
   ['2810.17', '0.000'],
   ['2810.18', '0.000'],
   ['2810.19', '0.008'],
   ['2810.20', '0.015'],
   ['2810.22', '0.074'],
   ['2810.25', '6.027'],
   ['2810.27', '0.356'],
   ['2810.29', '1.068'],
   ['2810.30', '27.383'],
   ['2810.31', '12.261'],
   ['2810.32', '0.015'],
   ['2810.33', '0.376'],
   ['2810.34', '13.115'],
   ['2810.36', '5.335'],
   ['2810.37', '1.057'],
   ['2810.38', '2.669'],
   ['2810.39', '0.008'],
   ['2810.40', '4.046'],
   ['2810.41', '0.020'],
   ['2810.42', '15.359'],
   ['2810.43', '0.000'],
   ['2810.46', '0.356'],
   ['2810.47', '2.866'],
   ['2810.48', '3.806'],
   ['2810.49', '11.009'],
   ['2810.50', '6.180'],
   ['2810.51', '0.074'],
   ['2810.52', '0.122'],
   ['2810.53', '0.956'],
   ['2810.54', '14.644'],
   ['2810.55', '0.066'],
   ['2810.56', '0.371'],
   ['2810.57', '22.166'],
   ['2810.58', '0.194'],
   ['2810.59', '18.120'],
   ['2810.60', '3.685'],
   ['2810.61', '17.349'],
   ['2810.62', '31.983'],
   ['2810.63', '5.032'],
   ['2810.64', '13.578'],
   ['2810.65', '0.075'],
   ['2810.66', '0.000'],
   ['2810.67', '21.331'],
   ['2810.68', '0.015'],
   ['2810.72', '14.219'],
   ['2810.73', '5.175'],
   ['2810.74', '0.356'],
   ['2810.79', '3.915'],
   ['2810.80', '18.246'],
   ['2810.82', '0.107'],
   ['2810.83', '6.638'],
   ['2810.84', '19.753'],
   ['2810.87', '14.738'],
   ['2810.88', '22.781'],
   ['2810.90', '17.987'],
   ['2810.93', '19.608'],
   ['2810.95', '0.928'],
   ['2810.97', '32.456'],
   ['2810.98', '0.347'],
   ['2810.99', '8.034'],
   ['2811.03', '7.844'],
   ['2811.06', '5.057'],
   ['2811.09', '47.085'],
   ['2811.10', '29.136'],
   ['2811.14', '14.229'],
   ['2811.16', '0.557'],
   ['2811.17', '2.312'],
   ['2811.18', '9.130'],
   ['2811.20', '11.660'],
   ['2811.21', '2.504'],
   ['2811.22', '8.693'],
   ['2811.24', '11.847'],
   ['2811.27', '0.000'],
   ['2811.29', '0.445'],
   ['2811.32', '0.274'],
   ['2811.33', '0.673'],
   ['2811.35', '0.000'],
   ['2811.36', '0.028'],
   ['2811.40', '26.592'],
   ['2811.47', '4.275'],
   ['2811.50', '0.000'],
   ['2811.52', '25.015'],
   ['2811.54', '0.000'],
   ['2811.61', '0.124'],
   ['2811.63', '0.017'],
   ['2811.65', '7.355'],
   ['2811.66', '0.235'],
   ['2811.67', '0.113'],
   ['2811.68', '46.582'],
   ['2811.72', '0.829'],
   ['2811.73', '8.261'],
   ['2811.75', '0.311'],
   ['2811.77', '15.679'],
   ['2811.78', '208.370'],
   ['2811.79', '10.576'],
   ['2811.84', '32.601'],
   ['2811.89', '0.794'],
   ['2812.04', '0.054'],
   ['2812.05', '0.311'],
   ['2812.06', '0.534'],
   ['2812.08', '1.126'],
   ['2812.11', '5.714'],
   ['2812.15', '1.335'],
   ['2812.18', '1.259'],
   ['2812.19', '0.000'],
   ['2812.21', '0.702'],
   ['2812.24', '0.015'],
   ['2812.25', '14.158'],
   ['2812.26', '1.837'],
   ['2812.27', '0.109'],
   ['2812.28', '0.015'],
   ['2812.29', '1.490'],
   ['2812.30', '0.000'],
   ['2812.34', '47.737'],
   ['2812.40', '0.315'],
   ['2812.41', '0.320'],
   ['2812.43', '0.611'],
   ['2812.45', '72.485'],
   ['2812.48', '1.134'],
   ['2812.59', '1.744'],
   ['2812.64', '34.417'],
   ['2812.69', '15.366'],
   ['2812.82', '4.055'],
   ['2812.89', '0.000'],
   ['2813.01', '7.396'],
   ['2813.04', '3.911'],
   ['2813.08', '0.015'],
   ['2813.11', '3.396'],
   ['2813.18', '76.323'],
   ['2813.37', '44.998'],
   ['2813.67', '12.779'],
   ['2813.93', '0.112'],
   ['2814.14', '28.572'],
   ['2814.50', '4.550'],
   ['2814.95', '0.000'],
   ['2814.97', '1.335'],
   ['2815.09', '1.534'],
   ['2815.16', '2.028'],
   ['2815.41', '0.533'],
   ['2815.43', '0.020'],
   ['2816.08', '60.142'],
   ['2816.27', '36.296'],
   ['2816.56', '44.879'],
   ['2818.69', '0.160'],
   ['2818.71', '3.769'],
   ['2820.88', '0.718'],
   ['2821.98', '0.319'],
   ['2822.18', '25.628'],
   ['2822.44', '0.011'],
   ['2822.46', '40.465'],
   ['2823.29', '25.000'],
   ['2823.97', '0.000'],
   ['2826.94', '0.598'],
   ['2839.26', '1.000'],
   ['2839.27', '0.024'],
   ['2839.28', '0.360'],
   ['2861.01', '11.560'],
   ['2861.02', '11.724'],
   ['2866.02', '7.076'],
   ['2869.59', '0.009'],
   ['2869.65', '9.762'],
   ['2869.93', '0.000'],
   ['2950.31', '0.000']]
}
```
Could you please help me understand this behavior? Thank you in advance for your help

### Code

```
  

```

**Comment by pcriadoperez** [comment]
Hi @jsmerix , thanks a lot for reporting and being so detailed.
Can you provide what you used for the options of 'fetchBalanceSnapshot' and 'awaitBalanceSnapshot' ?

**Comment by jsmerix** [comment]
We haven't explicitly altered these values, checking out the default options for binance in `4.2.45`, it says:
`{'fetchBalanceSnapshot': False, 'awaitBalanceSnapshot': True}`

**Comment by aifirstd3v** [comment]
I'm curious if this issue is fixed.

**Comment by samgermain** [comment]
@carlosmiei Can you take a look at this? By looking at `binance.handleOrderBook`, it looks like the problem is caused by something to do with the value stored inside `orderbook['nonce']`, I think that one of the instances must have a nonce value that is more up to date than the other. It looks like the second data point `asks` has some values excluded from it which are present in the first data point

**Comment by xalumok** [comment]
Are there any updates on this issue?

**Comment by pcriadoperez** [comment]
Hi, I ran the following script to find differences between two instances. Here were my conclusions:
 - I'm able to reproduce the behavior @jsmerix  reports
 - The bids/asks that are different come from a difference between the snapshot received from binance and the snapshot received a couple nonces back and built using the depth update messages
 - This only happens in the bids/asks with lowest/highest price (least significant)
 - Binance admits that there can be slight differences between the snapshot and the built order book (see below from their docs)
 - In the case of spot markets you can improve by setting `this.options['watchOrderBookLimit'] = 5000` and calling watchOrderBook with a limit of 5000, so that the outliers are even less significant.
 
 Note from binance:
Due to depth snapshots having a limit on the number of price levels, a price level outside of the initial snapshot that doesn't have a quantity change won't have an update in the Diff. Depth Stream. Consequently, those price levels will not be visible in the local order book even when applying all updates from the Diff. Depth Stream correctly and cause the local order book to have some slight differences with the real order book. However, for most use cases the depth limit of 5000 is enough to understand the market and trade effectively.


- Bash script used for testing: I would compare after in VScode instead of the diff file.
```bash
##!/bin/bash

# Run the command twice concurrently and save the outputs to different files
command="tsx examples/ts/cli.ts binance watchOrderBook BTC/USDT --verbose"

$command > log1.txt &
pid1=$!

$command > log2.txt &
pid2=$!

# Wait for 20 seconds
sleep 10

# Kill both processes
kill $pid1 $pid2

# Give the processes a moment to terminate and capture their output
wait $pid1 2>/dev/null
wait $pid2 2>/dev/null

# Compare the two output files
diff log1.txt log2.txt > diff.txt

# Display the differences (if any)
if [ -s diff.txt ]; then
    echo "Differences found between log1.txt and log2.txt:"
    cat diff.txt
else
    echo "No differences found between log1.txt and log2.txt."
fi

```

In conclusion I don't think this is a ccxt issue. I'll close this issue for now but feel free to reopen or reach out with any comments
